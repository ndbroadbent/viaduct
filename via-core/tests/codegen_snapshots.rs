use std::path::Path;

use anyhow::Result;
use via_core::{codegen, parser};

#[test]
fn generates_expected_outputs_for_article_fixture() -> Result<()> {
    let fixture = Path::new("tests/fixtures/article.via");
    let resources = parser::parse_file(fixture)?;
    assert_eq!(resources.len(), 1);

    let generation = codegen::generate(&resources)?;

    let mut files = generation.files;
    files.sort_by(|a, b| a.relative_path.cmp(&b.relative_path));

    for file in files {
        let path_str = file.relative_path.to_string_lossy().replace('/', "__");
        let snapshot_name = format!("article__{}", path_str);
        insta::assert_snapshot!(snapshot_name, file.contents);
    }

    Ok(())
}
