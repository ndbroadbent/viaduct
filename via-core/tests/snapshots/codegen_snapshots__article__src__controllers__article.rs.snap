---
source: tests/codegen_snapshots.rs
expression: file.contents
---
// @generated by via-core. DO NOT EDIT BY HAND.
// source: tests/fixtures/article.via

use loco_rs::prelude::*;
use serde_json::json;
use crate::models::{ArticleCreateParams, ArticleUpdateParams};

pub const RESPOND_WITH: &[&str] = &["json"];

pub fn routes() -> Routes {
    Routes::new()
        .prefix("articles/")
        .add("/", get(index))
        .add("/{id}", get(show))
        .add("/", post(create))
        .add("/{id}", put(update))
        .add("/{id}", delete(destroy))
}

pub async fn index(State(_ctx): State<AppContext>) -> Result<Response> {
    format::json(json!({"todo": "Article#index"}))
}

pub async fn show(State(_ctx): State<AppContext>, Path(id): Path<String>) -> Result<Response> {
    format::json(json!({"todo": "Article#show", "id": id}))
}

pub async fn create(State(_ctx): State<AppContext>, Json(payload): Json<ArticleCreateParams>) -> Result<Response> {
    format::json(json!({"todo": "Article#create", "payload": payload}))
}

pub async fn update(State(_ctx): State<AppContext>, Path(id): Path<String>, Json(payload): Json<ArticleUpdateParams>) -> Result<Response> {
    format::json(json!({"todo": "Article#update", "id": id, "payload": payload}))
}

pub async fn destroy(State(_ctx): State<AppContext>, Path(id): Path<String>) -> Result<Response> {
    format::json(json!({"todo": "Article#destroy", "id": id}))
}

// Associated model fields
// - title: String
// - body: Option<String>
// - published_at: Option<String> (ISO 8601 DateTime string; swap for chrono if desired)
// - author_name: String
